text/0031-corejam.md

```diff
@@ -0,0 +1,644 @@
+
+Finally, at some point later still `T+r+i+a+o`, the Results of the Work Package are aggregated into groups of Work Classes, and then *Pruned* and *Accumulated* into the common state of the Relay-chain. This process is known as *Integration* (in the fixed-function parachains model, this is known as "inclusion") and is irreversible within any given fork. Additional latency from being made *Available* to being *Integrated* (i.e. the `a` component) may be incurred due to ordering requirements, though it is expected to be zero in the variant of this proposal to be implemented initially (see **Work Package Ordering**, later).
+
+### Collect-Refine
+
+The first two stages of the CoreJam process are *Collect* and *Refine*. *Collect* refers to the collection and authorization of Work Packages (collections of items together with an authorization) to utilize a Polkadot Core. *Refine* refers to the performance of computation according to the Work Packages in order to yield a *Work Result*. Finally, each Validator Group member attests to a Work Package yielding a set of Work Results and these Attestations form the basis for bringing the Results on-chain and integrating them into the Polkadot (and in particular the Work Class's) state which happens in the following stages.
+
+#### Collection and `is_authorized`
+
+Collection is the means of a Validator Group member attaining a Work Package which is authorized to be performed on their assigned Core at the current time. Authorization is a prerequisite for a Work Package to be included on-chain. Computation of Work Packages which are not Authorized is not rewarded. Incorrectly attesting that a Work Package is authorized is a disputable offence and can result in substantial punishment.
+
+There are two kinds of Authorization corresponding to the two kinds of Coretime which are sold by Polkadot (see RFC#0001). An Authorization for usage of Instantaneous Coretime consists of a self-contained Signature of an account which own enough Instantaneous Coretime Credit in order to purchase a block of Coretime at the current rate signing a payload of the Work Package hash.
+
+RcVGs run the risk of a credit owner not having the credit at the point of eventual payment (in the Join stage, later), in which case the RcVG will not be rewarded. Credit may never be withdrawn, therefore RcVGs can safely accept a block if and only if the Credit account contains a balance of at least the product of the number of Cores assigned to IC, the price per IC core per block and the number of blocks behind the head of the finalized chain which the RcVG currently may be.
```

Comment:
> As per RFC#1 IC is bought before building and sending the work package. Why change this? 

RFC#1 states that IC *credits* are purchased on the Broker chain. These credits are assigned to a collator account. They may then be spent on the Relay-chain chain by that collator at the point that the collator provides a parachain block.

Aside from "collator"/"parachain block" becoming "builder"/"work package", nothing changes.

Reply:
> InstantaneousAuth could still be quite light weight, by it being a signature over a bought IC.

Yes, it would be my expectation that `InstantaneousAuth` is a Signature from of an account which holds IC credits (on the Relay-chain).

Reply:
> that collator at the point that the collator provides a parachain block.

But the collator doesn't know to whom to provide the parachain block. In the current model the collators know the validators before. Need to open a connection and advertise their collation. If the collator doesn't know to whom to send a collation, it could also mean that we overload a group because multiple collators try to send them the data. It is also less wasted work, if you first try to acquire an IC slot and then build a block.

Reply:
Sure this would need to be solved. However this presupposed "solution" of hammering the Relay-chain with ICT purchase-and-schedule transactions does not scale.

One possible alternative would be to assign a core randomly on the basis of a VRF from the pool of ICT cores. Assuming that this pool does not change too often (given RFC-1 it would only change every 80 blocks), then this should work ok.

Reply:
> This seems exceptionally specialised to collators and cumulus. Having read it I've still no idea how it relates to this RFC. Perhaps you could provide the general gist here?

It describes how to separate orders of cores from actually authoring blocks for them, while maintaining desired properties.  The described issues do not exist in that form, if we go with the approach of this RFC - where the payment for the core happens at the same time the collation/work package is provided.

> It would be possible to use the same Authorization framework for ICT as well as BCT. However, it comes at a cost: whereas with the current proposal ICT purchasing happens implicitly as a function of the Work Package, with the Authorization framework the purchase and Work Package submission would need to be two separate transactions with, presumably, some cost in terms of latency, state I/O, processing, scheduling and queuing.

Indeed, there is some additional latency - which we considered an acceptable trade-off and if need be, it could be reduced to barely [any](https://github.com/paritytech/polkadot-sdk/issues/1312). The question is, how instant does instantaneous core time need to be? If for example 2 blocks of latency are acceptable, we could move ordering extrinsics to the core-time chain and transmit checked orders in batches to the relay chain.

With regards to scheduling and queuing: Some form of scheduling and queuing will be required in any approach. In the model described in this RFC, it is not clear at all to which backing group a given work package should be provided to.

> These circumstances are antithetical to those of ICT. We know exactly which Builder is building the block since they are the ones who purchased this Coretime. The question of whether the Builder (and their Work Package) is in alignment with the Coretime purchaser is therefore moot as they are one and the same economic force. As long as the Coretime is paid for (something which we can statically prove), then the Work Package should be allowed to run.

Precisely. The described problem does not exist in this RFC, but it faces other challenges. An important one: Yes it should be allowed to run, but where? - The actual scheduling is not answered. 

> Sure this would need to be solved. However this presupposed "solution" of hammering the Relay-chain with ICT purchase-and-schedule transactions reduces scalability.

As mentioned, depending on what latency needs we actually envision, those orders would not necessarily need to go to the relay chain.

> One possible alternative would be to assign a core randomly on the basis of a VRF from the pool of ICT cores. Assuming that this pool does not change too often (given RFC-1 it would only change every 80 blocks), then this should work ok.

The way it is implemented right now, just aligns very well with how scheduling is implemented already. @rphmeier already explored different designs with RFC-3. Given that CoreJam is a huge change anyway, we can certainly re-evaluate designs here. The UX/raciness concern remains though. All in all, it feels like handling on-demand mid- to long term might be an entirely different discussion*) anyway. Maybe we leave exact details open as of this RFC and just state desired properties?


*) E.g. it is not clear to me, whether on-demand/instantaneous  on the relay chain will even need to exist with generalized builders. If we authenticate builders not parachains, builders could just order bulk and then provide an instantaneous service themselves using their bulk block space. Which would be another elegant way of getting rid of user-provided extrinsics on the relay chain.

Reply:
> Assuming that this pool does not change too often (given RFC-1 it would only change every 80 blocks), then this should work ok.

Actually, the pool as in cores available for instantaneous, can change each block with RFC-1. As only a fraction of those 80 blocks might be dedicated to instantaneous.

Reply:
> Actually, the pool as in cores available for instantaneous, can change each block with RFC-1. As only a fraction of those 80 blocks might be dedicated to instantaneous.

RFC-3 does indeed allow for changes to the core assignment on a per-block basis. However RFC-1 is structured so that these changes can only happen on timeslice boundaries (every 80 blocks). Within a timeslice, there's a known amount of coretime given over to instantaneous on a known subset of cores. By weighting a random distribution according to this coretime-core distribution we can get a reasonably fair core assignment.